### ISSUE TITLE
Use branch names to determine version bump

### ISSUE THREAD

**Issue by sean-abbott** [pypa/setuptools-scm#75]
In order to fit seamlessly with a developers workflow,setuptools_scm should be (optionally) branch aware, and be able to use the branch prefix in order to determine which version part to bump.

Which branch prefixes equate to which version parts should be configurable.  For a sane default, I'm thinking:

no prefix, unrecognized prefix, or bugfix/ == patch version bump (i.e., if master is tagged 1.0.0, and we branch bugfix/ISSUE-10), then the automatic version generation on that branch would become 1.0.1.dev{distance from branch}.n{revision hash}

feature/ == minor version bump (i.e., if master is tagged 1.0.0, and we branch feature/ISSUE-10, then the automatic version on that branch would be 1.1.0.dev{distance from branch}.n{revision hash}

In the configuration file, you might need to add configuration options for each version part.  i.e.

patch_branch: comma separated list of branch prefixes that denote a patch bump
minor_branch: comma separated list of branch prefixes that denote a minor bump
major_branch: comma separated list of branch prefixes that denote a major version bump
(and this config might also want a special option to denote "any unrecognized branch prefix" and "no prefix")

The one piece of this I haven't entirely figured out yet is releasing.  Either you might need a new setup.py action (i.e., python setup.py release) that performs similarly) to zest pre/release/post, but can handle the git merge and tag, or perhaps in order to use this feature you MUST use the version file so that when you merge you'll get the proper tag.  Haven't quite figured that out yet.

This issue is being raised as a result of discussion at zest releaser:  https://github.com/zestsoftware/zest.releaser/issues/121#issuecomment-171089841

Please let me know if you need more detail.

**Comment by jaraco** [comment]
I'm generally in favor of this suggestion, especially given the wide adoption of "flow" (in Git and Mercurial), but even more in that the defaults are configurable to work with other branching methodologies. My preference would be for the default bump to be 0.1 (as far too many non-bugfix changes are released with an 0.0.1 bump), but I acknowledge that many projects would prefer the behavior as specified above, and that's probably more consistent with the existing behavior.

I haven't considered yet the deeper implications, but I wanted to first say thanks for laying out a basic spec and endorse the idea.

**Comment by RonnyPfannschmidt** [comment]
basically this entails a few new concepts
- configurable guessed bump granularity
- creating tags to bump/release

as for configurable bump granularity,
my suggestion would be to map regex patterns of branch names to increments

something like 

``` python
def guess_next_version(version):
    from setuptools_scm.bumping import (
        branch_bump,
        guess_next_dev,
        larger_of,
        semver,
    )
    return larger_of(
       branch_bump(version, [
          ('feature/.*', semver.minor),
          ('hotfix/.*, semver.patch),
       ]),
       guess_next_dev(version),
      )
```

comes to mind,
a few simple variations could be shipped

for bump commands, i see a need for 2 commands
one to bump development versions and one to tag a release

as for flow - that model is totally broken wrt obtaining sane metadata from the scm

**Comment by sean-abbott** [comment]
I'm guessing that if I try and write a plugin for this, I should not use non-core libraries, like semanticversion?

**Comment by RonnyPfannschmidt** [comment]
Im not sure a Plugin can easily provide the entry points and I suspect extra dependencies created trouble until the build tool pep will land

I proposed a brainstorming/discussion session before attempting

**Comment by RonnyPfannschmidt** [comment]
@sean-abbott - with recent enhancements, this is the next thing i#d like to implement, are you up for a design discussion/brainstorming?

also a ping to @jaraco wrt the same

**Comment by sean-abbott** [comment]
I'd be in to the discussion, but tbh I've completely forgotten what I was doing and I no longer work for the company where I had the usecase. :-/

**Comment by sean-abbott** [comment]
Also, sorry I missed this until 6 months later.  Dangit.

**Comment by jaraco** [comment]
I honestly don't have the bandwidth to provide meaningful insight here. I'm only barely familiar enough with Git flow that I use it sparingly and sometimes incorrectly. I think we'd need some input from someone with a deeper understanding of Flow and use-cases to design and test against. I'm happy to review an implementation and test it, though.

**Comment by RonnyPfannschmidt** [comment]
thanks for the inputs, i plan to provide the branch name in the metadata and let users decide by version scheme, however on as i have time to work on it basis

**Comment by sean-abbott** [comment]
I am also happy to test, and I do have at least one project where I can do so meaningfully.

**Comment by RonnyPfannschmidt** [comment]
ref #240

**Comment by RonnyPfannschmidt** [comment]
fixed by #240
